### ISSUE TITLE
Missing BEGIN seems to cause out of sync sessions.

### ISSUE THREAD

**Issue by cdcformatc** [pallets/flask-sqlalchemy#631]
Background: Apache server using mod_wsgi to serve a Flask app using Flask_Sqlalchemy connecting to MySQL. This is a full stack application so it is nearly impossible to create a minimal example but I have tried.

My problem is that when I make some change that should modify the database subsequent requests don't always seem to reflect that change. For example if I create an object, then try to edit that same object, the edit will sometimes fail.

When turning on SQLALCHEMY_ECHO and SQLALCHEMY_COMMIT_ON_TEARDOWN I can see that when the response is good everything is framed by a BEGIN and a COMMIT. But when the response is bad, these things are missing.

Example code:

    bp = Blueprint('api_region', __name__, url_prefix='/app/region')
    @bp.route('/rename/<int:region_id>/<string:name>', methods=['POST'])
    def change_name(region_id, name):
        region = Region.query.get(region_id)
        try:
            region.name = name
        except AttributeError:
            abort(404)
        db.session.add(region)
        db.session.commit()
        return "Success"

    @bp.route('/name/<int:region_id>/', methods=['GET'])
    def get_name(region_id):
        print("--- Get Start--- ")
        region = Region.query.get(region_id)
        try:
            name = region.name
        except AttributeError:
            abort(404)
        print("--- Get End ---")
        return name

After object is created send a POST

    curl -X POST https://example.com/app/region/rename/5/Europe
Then several GETs

    curl -X GET https://example.com/app/region/name/5/

Sometimes, the GET will return the correct info, but every now and then it will return whatever it was before. 
Example Output:

    format@pixel:~$ curl -X POST https://example.com/app/region/rename/5/@@OLD@@
    Success
    format@pixel:~$ curl -X POST https://example.com/app/region/rename/5/New
    Success
    format@pixel:~$ for i in {1..25}; do  echo ; curl -X GET https://example.com/app/region/name/5/; done
    New
    New
    New
    New
    New
    New
    @@OLD@@
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New
    New

This first log snippet is typical for when the request is correct. You can see the implicit BEGIN and the COMMIT on teardown. 
  
    [Tue Jul 24 17:49:23.217666 2018] [wsgi:error] [pid 27919:tid 139730801542912] --- Get Start ---
    [Tue Jul 24 17:49:23.219964 2018] [wsgi:error] [pid 27919:tid 139730801542912] 2018-07-24 17:49:23,219 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
    [Tue Jul 24 17:49:23.220053 2018] [wsgi:error] [pid 27919:tid 139730801542912] INFO:sqlalchemy.engine.base.Engine:BEGIN (implicit)
    [Tue Jul 24 17:49:23.221282 2018] [wsgi:error] [pid 27919:tid 139730801542912] 2018-07-24 17:49:23,221 INFO sqlalchemy.engine.base.Engine SELECT region.id AS region_id, region.name AS region_name, region.north AS region_north, region.west AS region_west, region.south AS region_south, region.east AS region_east, region.coords AS region_coords 
    [Tue Jul 24 17:49:23.221321 2018] [wsgi:error] [pid 27919:tid 139730801542912] FROM region 
    [Tue Jul 24 17:49:23.221336 2018] [wsgi:error] [pid 27919:tid 139730801542912] WHERE region.id = %s
    [Tue Jul 24 17:49:23.221356 2018] [wsgi:error] [pid 27919:tid 139730801542912] 
    [Tue Jul 24 17:49:23.221414 2018] [wsgi:error] [pid 27919:tid 139730801542912] INFO:sqlalchemy.engine.base.Engine:SELECT region.id AS region_id, region.name AS region_name, region.north AS region_north, region.west AS region_west, region.south AS region_south, region.east AS region_east, region.coords AS region_coords 
    [Tue Jul 24 17:49:23.221433 2018] [wsgi:error] [pid 27919:tid 139730801542912] FROM region 
    [Tue Jul 24 17:49:23.221447 2018] [wsgi:error] [pid 27919:tid 139730801542912] WHERE region.id = %s
    [Tue Jul 24 17:49:23.221465 2018] [wsgi:error] [pid 27919:tid 139730801542912] 
    [Tue Jul 24 17:49:23.221683 2018] [wsgi:error] [pid 27919:tid 139730801542912] 2018-07-24 17:49:23,221 INFO sqlalchemy.engine.base.Engine (5,)
    [Tue Jul 24 17:49:23.221767 2018] [wsgi:error] [pid 27919:tid 139730801542912] INFO:sqlalchemy.engine.base.Engine:(5,)
    [Tue Jul 24 17:49:23.222922 2018] [wsgi:error] [pid 27919:tid 139730801542912] 2018-07-24 17:49:23.222868 New
    [Tue Jul 24 17:49:23.222990 2018] [wsgi:error] [pid 27919:tid 139730801542912] --- Get End ---
    [Tue Jul 24 17:49:23.224867 2018] [wsgi:error] [pid 27919:tid 139730801542912] 2018-07-24 17:49:23,224 INFO sqlalchemy.engine.base.Engine COMMIT
    [Tue Jul 24 17:49:23.224946 2018] [wsgi:error] [pid 27919:tid 139730801542912] INFO:sqlalchemy.engine.base.Engine:COMMIT
    
This next snippet is typical for when the request is incorrect. You can see it is performing the SELECT the same as before but, there is no BEGIN nor a COMMIT. Not sure what this means or why it occurs, but it leads me to believe the problem is in the SQLAlchemy or MySQL layer. 

    [Tue Jul 24 17:49:23.288116 2018] [wsgi:error] [pid 27919:tid 139730654033664] --- Get Start ---
    [Tue Jul 24 17:49:23.290049 2018] [wsgi:error] [pid 27919:tid 139730654033664] 2018-07-24 17:49:23,289 INFO sqlalchemy.engine.base.Engine SELECT region.id AS region_id, region.name AS region_name, region.north AS region_north, region.west AS region_west, region.south AS region_south, region.east AS region_east, region.coords AS region_coords 
    [Tue Jul 24 17:49:23.290079 2018] [wsgi:error] [pid 27919:tid 139730654033664] FROM region 
    [Tue Jul 24 17:49:23.290092 2018] [wsgi:error] [pid 27919:tid 139730654033664] WHERE region.id = %s
    [Tue Jul 24 17:49:23.290109 2018] [wsgi:error] [pid 27919:tid 139730654033664] 
    [Tue Jul 24 17:49:23.290168 2018] [wsgi:error] [pid 27919:tid 139730654033664] INFO:sqlalchemy.engine.base.Engine:SELECT region.id AS region_id, region.name AS region_name, region.north AS region_north, region.west AS region_west, region.south AS region_south, region.east AS region_east, region.coords AS region_coords 
    [Tue Jul 24 17:49:23.290186 2018] [wsgi:error] [pid 27919:tid 139730654033664] FROM region 
    [Tue Jul 24 17:49:23.290197 2018] [wsgi:error] [pid 27919:tid 139730654033664] WHERE region.id = %s
    [Tue Jul 24 17:49:23.290214 2018] [wsgi:error] [pid 27919:tid 139730654033664] 
    [Tue Jul 24 17:49:23.290422 2018] [wsgi:error] [pid 27919:tid 139730654033664] 2018-07-24 17:49:23,290 INFO sqlalchemy.engine.base.Engine (5,)
    [Tue Jul 24 17:49:23.290499 2018] [wsgi:error] [pid 27919:tid 139730654033664] INFO:sqlalchemy.engine.base.Engine:(5,)
    [Tue Jul 24 17:49:23.291567 2018] [wsgi:error] [pid 27919:tid 139730654033664] 2018-07-24 17:49:23.291529 @@OLD@@
    [Tue Jul 24 17:49:23.291591 2018] [wsgi:error] [pid 27919:tid 139730654033664] --- Get End ---

**Comment by rsyring** [comment]
If you believe this is a bug in FSA, we are going to need you to create a [MCVE](https://stackoverflow.com/help/mcve) of the problem you are having.  If you'd like to take a look at how to do that, the tests of this project demonstrate how to create very minimal flask apps for testing purposes.

**Comment by cdcformatc** [comment]
I had solved the problem I am not sure if it is a bug in FSA. For future reference the issue disappeared when I removed the line `app.app_context().push()` from my application factory used for unit testing purposes. I am not sure how pushing a context like this changes the behavior of FSA and only some of the time, but the line was leftover from very early days in the project based off a flask tutorial.
